### PR TITLE
openPMD: Expose ADIOS2 Operators (Compressors)

### PR DESCRIPTION
**Pull request by ax3l**
Expose ADIOS2 Operators (at the moment: one operator max) in WarpX inputs. This allows the user to select ZFP or Blosc (e.g. with ZSTD) compression without changing the input file.

We can later on define short-hand notions for this. For now we give the parameters control out for the expert user and document good examples.

## Background

This provides functionality as in our ISC17 / DRBSD-1 paper: https://arxiv.org/abs/1706.00522

## Tests

In my tests, the documented options work and make a difference. I am not sure if the blosc threading works (problem noted, debugged and moved upstream) - if that is a problem it will be fixed with a later release of ADIOS2.

### CODE REVIEW

File: Source/Diagnostics/WarpXOpenPMD.H [hunk 1]
```diff
@@ -245,6 +251,7 @@ private:
 
   bool m_OneFilePerTS = true;  //! write in openPMD fileBased manner for individual time steps
   std::string m_OpenPMDFileType = "bp"; //! MPI-parallel openPMD backend: bp or h5
+  std::string m_OpenPMPoptions = "{}"; //! JSON option string for openPMD::Series constructor
```
Comment (guj) [inline Source/Diagnostics/WarpXOpenPMD.H]: why m_OpenPMPoption, instead of m_OpenPMDOption?  Anything special with PMP?

**Review by guj — COMMENTED** [review]
Are the operator options expect to work for all file formats?

**Comment by ax3l** [comment]
> Are the operator options expect to work for all file formats?

No, at the moment only for ADIOS2. We need to add the compressor options to HDF5 first, as they now also work with parallel HDF5 these days :)

File: Source/Diagnostics/WarpXOpenPMD.H (hunk 1 above)
Comment (ax3l) [reply Source/Diagnostics/WarpXOpenPMD.H]: Oh no, that's a typo. Good catch, thanks!

**Comment by guj** [comment]
> > Are the operator options expect to work for all file formats?
> 
> No, at the moment only for ADIOS2. We need to add the compressor options to HDF5 first, as they now also work with parallel HDF5 these days :)

Hmm, if HDF5 is expecting not the same set of options,  would it be clear to name these operators adios_operator instead of openpmd_operator? 
from the face value,  a person would not have guessed that file format needs to be ADIOS in order for these operators to take place. 

further, will this format make sense in the future? <diag>.openpmd.<filetype><key><parameter_name>=<value>

<filetype>: bp/h5 etc
<key>: operator/transport/...  specific to <filetype>
<parameter_name><value>: pass to the openpmd-api to the <filetype> IO as is through  json,  like what you did.

**Comment by ax3l** [comment]
Thanks a lot! Updated accordingly :)

We can add an additional low-level `openpmd_options = "{some_json}"` option in a follow-up PR that can be used for any option.

And long-term we will add high-level flags that control multiple backends.

**Review by guj — APPROVED** [review]
Looks nice!

**Comment by guj** [comment]
> Update accordingly.
> 
> We can add an additional low-level `openpmd_options = "{some_json}"` option in a follow-up PR that can be used for any option.

That would be great as long as it will not make the input file look bulky :) maybe redirect to a json file instead of having json content as value.
> 
> And long-term we will add high-level flags that control multiple backends.

**Review by guj — APPROVED** [review]
Can merge

**Review by RemiLehe — APPROVED** [review]

**Pull request merged**